### ISSUE TITLE
Dockerize the build

### ISSUE THREAD

**Issue by asbjornu** [schismtracker/schismtracker#146]
As mentioned in [this tweet](https://twitter.com/womenwhogo_chi/status/956650375574106112) I thought it would be a good idea to build Schism Tracker with Docker so all dependencies are in place without having to install anything other than Docker on the local computer. Thoughts?

**Comment by polyzium** [comment]
I am personally never tried docker so i can't say anything about that. Though, some people prefer the traditional way of compiling.

**Comment by jangler** [comment]
Docker itself is way bigger than all Schism's dependencies, and there aren't many anyway… I personally don't see how this would be useful for more than a few people.

**Comment by asbjornu** [comment]
Using Docker has nothing to do with size, really. It's about not having to understand `make`, `autotools` and every infrastructural build knob Schism Tracker uses. It makes Schism Tracker easy to build and run for non-C developers (like myself).

**Comment by algorev** [comment]
Yes, but it makes docker, with its installation and root rights (at least on Mac) requirement the default. If you don't want to understand Make, then you probably don't need to build anything from source anyway. Plus, it makes the build process less portable. Finally, as a Mac user without root rights, it would prevent me from building Schism Tracker.

**Comment by asbjornu** [comment]
I've contributed a bit to Schism Tracker without being a C developer. I had never heard of autotools before I started poking at this project. There might be others like me who would like to contribute, but find the build setup so hard and daunting to get working that they don't. That's just a guess, though. I have no idea whether those people actually exist. :)

**Comment by ghost** [comment]
There are two types of FOSS enthusiast... Those who know how to use autotools, and those who are about to find out.

In seriousness, the Wikipedia article says that docker is "very limited" on Windows, which I'd be cautious about. Is it just a Unix thing?

Otherwise I have no objections to whatever, as long as it's simple, lightweight, not proprietary, and works on all the main platforms. Happy to test it if you want to put something together.

**Comment by asbjornu** [comment]
> There are two types of FOSS enthusiast... Those who know how to use autotools, and those who are about to find out.

😄 

> In seriousness, the Wikipedia article says that docker is "very limited" on Windows, which I'd be cautious about. Is it just a Unix thing?

In my experience, Docker works fine on Windows Server 2016 and Windows 10. I'll look into building an image when I find time.

**Comment by algorev** [comment]
I still think it would complicate things. Docker is complicated to install
and not everyone can install it. Plus, it is not available in most of the
systems where Schism Tracker can be built.

**Comment by asbjornu** [comment]
@algorev, I did not in any way intend for Docker to be the only way to build Schism Tracker. It's just an option for people who want to contribute without setting up an entire development environment.

**Comment by algorev** [comment]
Oh, my bad. I didn't understand it that way. Generally, when I see a build
system switch, the old one stops getting maintained.

**Comment by vcaputo** [comment]
Isn't this essentially just a proposal for adding some kind of machine-readable manifest to the source tree describing all the dependencies and how to construct a build host containing them?

I haven't been following docker developments for years now, but back in the day this was a Dockerfile.  There's been a lot of noise since then about standardization of containers and image formats, maybe it's still just a Dockerfile or maybe CNCF has ratified something more vendor-neutral since then.  There's also Flatpak which has its own manifest format.

I don't see a major problem with supporting some of these things, it shouldnt alter the build system at all.  This should be supplemental metadata describing the build environment needed for using the build system as-is.  The only valid reason I see to refuse or choose one vs. the other is if you have an opinion on which of these competing formats should "win", adding support for e.g. docker certainly gets it more entrenched in our world.

**Comment by asbjornu** [comment]
>Isn't this essentially just a proposal for adding some kind of machine-readable manifest to the source tree describing all the dependencies and how to construct a build host containing them?

Yep!

> I haven't been following docker developments for years now, but back in the day this was a Dockerfile. There's been a lot of noise since then about standardization of containers and image formats, maybe it's still just a Dockerfile or maybe CNCF has ratified something more vendor-neutral since then. There's also Flatpak which has its own manifest format.

While there are competing platforms, I'd say that for the atomic level of container technology, [Docker](https://www.docker.com/) is the winner. There are still lots of competition building on top of that for container orchestration, declarative syntaxes for building Docker images, etc., but it's hard to say which of them is going to come out on top yet.

> I don't see a major problem with supporting some of these things, it shouldnt alter the build system at all. This should be supplemental metadata describing the build environment needed for using the build system as-is. The only valid reason I see to refuse or choose one vs. the other is if you have an opinion on which of these competing formats should "win", adding support for e.g. docker certainly gets it more entrenched in our world.

Agreed!

**Comment by vcaputo** [comment]
@asbjornu What's preventing you from submitting a PR adding the appropriate Dockerfile and documentation on using it?

I don't think any of the devs are docker users, so it's unlikely they're going to put it together.

**Comment by asbjornu** [comment]
> What's preventing you from submitting a PR adding the appropriate Dockerfile and documentation on using it?

1. An agreement that this would be a useful feature at all. Spending time on a PR that will be declined is discouraging.
2. Time. 🕥

I'll submit a PR as soon as time allows. 😃

**Comment by jangler** [comment]
I would be OK with a Dockerfile PR. Having it as an option is mostly fine, but I imagine it shouldn't need more documentation than mentioning that the Dockerfile exists. People who already use Docker will know what to do, and people who don't use Docker can ignore it and follow the regular instructions.

**Comment by strongholdmedia** [comment]
Though I do find the notion of "not understanding Make means not building anything" somewhat harsh, I also do think that whatever is conceivable with docker is available to be used with lxc, with the sole exception of needing more knowledge than that so that script kiddies not rightclicking to flash around with "coding" and "building" skills.
(So no "prebuilt stuff").
Personally I think of this as a benefit, not a handicap - as long as the list of the dependencies is maintained and accessible.
